File: data_structures/linked_list/typescript/isPalindrome.ts
import ListNode from './SinglyLinkedListNode';

/**
 * Checks if given linked list is palindrome
 */
function isPalindrom<T>(head: ListNode<T>): boolean {
    if (!head || !head.next) {
        return true;
    }

    if (!head.next.next) {
        return head.val === head.next.val;
    }

    const stack = head.toArray();

    let i = 0;
    let j = stack.length - 1;

    while (i <= j) {
        if (stack[i] !== stack[j]) {
            return false;
        }

        i++;
        j--;
    }

    return true;
}

export default isPalindrom;

File: data_structures/linked_list/python/is_palindrome.py
from data_structures.linked_list.python.singly_linked_list_node import ListNode


def is_palindrome(head: ListNode) -> bool:
    """
    Checks if given LinkedList is palindrome
    """

    if not head or not head.next:
        return True

    if not head.next.next:
        return False

    stack = head.to_array()

    i = 0
    j = len(stack) - 1

    while i <= j:
        if stack[i] != stack[j]:
            return False
        i += 1
        j -= 1

    return True